src/cc/frontends/p4/compiler/ebpfProgram.py

```diff
@@ -0,0 +1,469 @@
+        self.entryPoints = []  # control-flow entry points from parser
+        self.counters = []
+        self.entryPointLabels = {}  # maps p4_node from entryPoints to labels in the C program
+        self.egressEntry = None
+
+        self.construct()
+
+        self.headersStructTypeName = "ebpf_headers_t"
+        self.headerStructName = "ebpf_headers"
+        self.metadataStructTypeName = "ebpf_metadata_t"
+        self.metadataStructName = "epbf_metadata"
+
+    def construct(self):
+        if len(self.hlir.p4_field_list_calculations) > 0:
+            raise NotSupportedException("{0} calculated field", self.hlir.p4_field_list_calculations.values()[0].name)
+
+        for h in self.hlir.p4_header_instances.values():
+            if h.max_index is not None:
+                assert isinstance(h, p4_header_instance)
+                if h.index == 0:
+                    # header stack; allocate only for zero-th index
+                    indexVarName = self.generateNewName(h.base_name + "_index")
+                    stack = ebpfInstance.EbpfHeaderStack(h, indexVarName, self.typeFactory)
+                    self.stacks.append(stack)
+            elif h.metadata:
+                metadata = ebpfInstance.EbpfMetadata(h, self.typeFactory)
+                self.metadata.append(metadata)
+            else:
+                header = ebpfInstance.EbpfHeader(h, self.typeFactory)
+                self.headers.append(header)
+
+        for p in self.hlir.p4_parse_states.values():
+            parser = ebpfParser.EbpfParser(p)
+            self.parsers.append(parser)
+
+        for a in self.hlir.p4_actions.values():
+            if self.isInternalAction(a):
+                continue
+            action = ebpfAction.EbpfAction(a, self)
+            self.actions.append(action)
+
+        for c in self.hlir.p4_counters.values():
+            counter = ebpfCounter.EbpfCounter(c, self)
+            self.counters.append(counter)
+
+        for t in self.hlir.p4_tables.values():
+            table = ebpfTable.EbpfTable(t, self, self.config)
+            self.tables.append(table)
+
+        for n in self.hlir.p4_ingress_ptr.keys():
+            self.entryPoints.append(n)
+
+        for n in self.hlir.p4_conditional_nodes.values():
+            conditional = ebpfConditional.EbpfConditional(n, self)
+            self.conditionals.append(conditional)
+
+        self.egressEntry = self.hlir.p4_egress_ptr
+
+    def isInternalAction(self, action):
+        # This is a heuristic really to guess which actions are built-in
+        # Unfortunately there seems to be no other way to do this
+        return action.lineno < 0
+
+    @staticmethod
+    def isArrayElementInstance(headerInstance):
+        assert isinstance(headerInstance, p4_header_instance)
+        return headerInstance.max_index is not None
+
+    def emitWarning(self, formatString, *message):
+        assert isinstance(formatString, str)
+        print "WARNING: ", formatString.format(*message)
```

Comment:
In order to be more python2/3 compatible, please use print() vs print. I typically use this as my reference: http://python-future.org/compatible_idioms.html